server/src/main/java/org/opensearch/wlm/QueryGroupService.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.wlm;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.cluster.metadata.QueryGroup;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.lifecycle.AbstractLifecycleComponent;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.monitor.jvm.JvmStats;
+import org.opensearch.monitor.process.ProcessProbe;
+import org.opensearch.search.ResourceType;
+import org.opensearch.search.backpressure.settings.NodeDuressSettings;
+import org.opensearch.search.backpressure.trackers.NodeDuressTrackers;
+import org.opensearch.wlm.cancellation.DefaultTaskCancellation;
+import org.opensearch.wlm.cancellation.DefaultTaskSelectionStrategy;
+import org.opensearch.wlm.tracker.QueryGroupResourceUsageTrackerService;
+import org.opensearch.threadpool.Scheduler;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.util.EnumMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Main service which will run periodically to track and cancel resource constraint violating tasks in QueryGroups
+ */
+public class QueryGroupService extends AbstractLifecycleComponent {
+    private static final Logger logger = LogManager.getLogger(QueryGroupService.class);
+
+    private final QueryGroupResourceUsageTrackerService queryGroupUsageTracker;
+    private final TimeValue queryGroupServiceRunIntervalInMillis;
+    private volatile Scheduler.Cancellable scheduledFuture;
+    private final ThreadPool threadPool;
```

Comment:
We should also have setting which controls whether the feature is enabled or not